File: HCISUPSDKV2.5.1.35_build20241101_Win64_ZH/demo/ISUPSDK_JAVA_DEMO_DeepInMind/src/DemoApp/SdkFunctionDemo.java
package DemoApp;

import Common.CommonMethod;
import SdkService.AlarmService.AlarmDemo;
import SdkService.CmsService.CmsDemo;
import SdkService.CmsService.HCISUPCMS;
import SdkService.SsService.SsDemo;
import SdkService.StreamService.StreamDemo;
import SdkService.StreamService.VoiceDemo;
import UIModule.PlaybackVideoUI;
import UIModule.PreviewVideoUI;

import java.io.UnsupportedEncodingException;

/**
 * @author zhengxiaohui
 * @date 2023/8/19 14:44
 * @desc Isup SDK 通用框架层的示例代码模块
 */
public class SdkFunctionDemo {

    public static void dispatch(String command, int lLoginID,
                                CmsDemo cmsDemo, AlarmDemo alarmDemo, SsDemo ssDemo, StreamDemo streamDemo, VoiceDemo voiceDemo) throws UnsupportedEncodingException, InterruptedException {
        switch (command) {
            /**
             * 约定： f0001 ~ f0999 预留给 [cms] 模块的代码示例
             */
            case "f0001": {
                System.out.println("\n[Function]CMS控制信令透传(需要设备在线, 获取NVR设备工作状态、远程升级等)");
                cmsDemo.CMS_XMLRemoteControl();
                break;
            }
            case "f0002": {
                System.out.println("\n[Function]ISUP透传接口(云存储配置))");
                cmsDemo.CMS_ISAPIPassThrough();
                break;
            }
            /**
             * 约定： f1000 ~ f1999 预留给 [ams] 模块的代码示例
             */
            /**
             * 约定： f2000 ~ f2999 预留给 [ss] 模块的代码示例
             */
            case "f2000": {
                System.out.println("\n[Function]上传图片至存储服务器(返回图片URL可以用于人脸下发)");
                //上传图片至存储服务器(返回图片URL可以用于人脸下发)
                String picName = CommonMethod.getConsoleInput("请输入想要上传的文件名称，且文件名带后缀(图片文件请提前放置在resources/pics 路径下)：" +
                        "\n如: 'FDLib.jpg' 具体请以您的实际图片名称为准!");

                String fileAbsPath = CommonMethod.getResFileAbsPath("pics/" + picName);
                ssDemo.ssCreateClient(fileAbsPath);
                ssDemo.ssUploadPic();
                ssDemo.ssDestroyClient();
                break;
            }
            /**
             * 约定： f3000 ~ f3999 预留给 [sms] 模块的代码示例
             */
            case "f3001": {
                /**
                 * 实时预览模块(需要设备在线, 需要实现前面初始化流媒体服务(StreamDemo.eStream_Init)的代码)
                 */
                System.out.println("\n[Function]取流预览模块(有预览窗口)");
                PreviewVideoUI.jRealWinInit();

                streamDemo.startRealPlayListen_Win();
                // FIXME 注意这里的IChannel，不同设备类型可能不太一样
                streamDemo.RealPlay(1);

                // 这里只预览20s, 方便demo示例代码的效果演示
                Thread.sleep(20 * 1000);

                streamDemo.StopRealPlay();

                //关闭预览窗口
                PreviewVideoUI.jRealWinDestroy();
                break;
            }
            case "f3002": {
                /**
                 * 实时预览模块(需要设备在线, 需要实现前面初始化流媒体服务(StreamDemo.eStream_Init)的代码)
                 */
                System.out.println("\n[Function]取流预览模块(保存到本地文件)");
                streamDemo.startRealPlayListen_File();
                // FIXME 注意这里的IChannel，不同设备类型可能不太一样
                streamDemo.RealPlay(1);

                // 这里只预览20s, 方便demo示例代码的效果演示
                Thread.sleep(20 * 1000);

                streamDemo.StopRealPlay();
                break;
            }
            case "f3003": {
                /**
                 * 按时间回放模块(需要设备在线, 需要实现前面初始化流媒体服务(StreamDemo.eStream_Init)的代码)
                 */
                System.out.println("\n[Function]按时间回放模块(有回放窗口)");
                PlaybackVideoUI.jRealWinInit();
                // 回放监听函数调用
                streamDemo.startPlayBackListen_WIN();
                // FIXME 注意这里的IChannel，不同设备类型可能不太一样
                // 开启回放预览
                streamDemo.PlayBackByTime(1);

                // 这里根据回调中收到的设备回放结束信令来判断设备推流结束，然后cms调用停止回放
                // 如果需要再回放过程中调用停止回放，则修改这里的while循环逻辑即可。
                while(!streamDemo.stopPlayBackFlag){

                    try {
                        Thread.sleep(1000); // 每秒检查一次
                    } catch (InterruptedException e) {
                        // 如果线程被中断，通常需要清除中断状态
                        Thread.currentThread().interrupt();
                        System.out.println("Playback was interrupted");
                        break; // 当前线程被中断时退出循环
                    }
                }

                //停止回放预览
                streamDemo.stopPlayBackByTime();

                //关闭预览窗口
                PlaybackVideoUI.jRealWinDestroy();
                break;
            }
            case "f3004": {
                /**
                 * 按时间回放模块(需要设备在线, 需要实现前面初始化流媒体服务(StreamDemo.eStream_Init)的代码)
                 */
                System.out.println("\n[Function]按时间回放模块(保存到本地文件)");
                // 回放监听函数调用
                streamDemo.startPlayBackListen_FILE();
                // FIXME 注意这里的IChannel，不同设备类型可能不太一样
                // 开启回放预览
                streamDemo.PlayBackByTime(1);

                // 这里根据回调中收到的设备回放结束信令来判断设备推流结束，然后cms调用停止回放
                // 如果需要再回放过程中调用停止回放，则修改这里的while循环逻辑即可。
                while(!streamDemo.stopPlayBackFlag){

                    try {
                        Thread.sleep(1000); // 每秒检查一次
                    } catch (InterruptedException e) {
                        // 如果线程被中断，通常需要清除中断状态
                        Thread.currentThread().interrupt();
                        System.out.println("Playback was interrupted");
                        break; // 当前线程被中断时退出循环
                    }
                }

                //停止回放预览
                streamDemo.stopPlayBackByTime();

                break;
            }
            case "f3005": {

                // 本demo音频编解码处理复用了设备网络sdk的音频编解码接口，所以需要导入设备网络sdk动态库到【/lib/netsdk】路径下进行加载，
                // 如果应用层自行编解码设备，则注释掉相关编解码代码即可。
                voiceDemo.voice_Init();
                voiceDemo.startVoiceServeListen();

                //获取设备通道对讲信息，包括编码格式，起始对讲通道号等，跟nvr对讲前先获取
                HCISUPCMS.NET_EHOME_DEVICE_INFO res = cmsDemo.getDevInfo();

                voiceDemo.makeVoice(20); //采集本地20s音频保存为pcm

                /**
                 * ISUP5.0语音转发模块(需要设备在线, 需要确定设备是否支持此功能, 需要实现前面初始化语音流媒体服务的代码)
                 */
                int dwVoiceChan = res.byStartDTalkChan + 3;
                byte dwAudioEncType = (byte) res.dwAudioEncType;

                voiceDemo.StartVoiceTrans(dwVoiceChan, dwAudioEncType);
                voiceDemo.StopVoiceTrans();
                break;
            }
            case "f3006":{

                //ISUP5.0抓图,ISUP抓图需要先开启存储服务，设置存储服务类型为云存储，下发云存储参数给到设备
                //抓图成功后，设备返回云存储URL，
                cmsDemo.GetPicByCloud(IsupTest.lLoginID);

            }
            case "f3007":{

                //云台转动
                cmsDemo.PTZCtrl(IsupTest.lLoginID);

            }
            case "f3008":{
                //录像查询
                cmsDemo.FindRecored(IsupTest.lLoginID);
            }
            case "f3009":{
                //获取设备信息
                cmsDemo.GetDevInfo(IsupTest.lLoginID);
            }

            case "f3010":{
                //手动校时
                cmsDemo.SetDevTimeCfg(IsupTest.lLoginID);
            }

            default: {
                System.out.println("未定义的指令类型!" + command);
            }
        }
    }
}

File: ISUPSDK_PYTHON_DEMO_BASE/src/DemoApp/SDKFunctionDemo.py
# --*-- conding:utf-8 --*--
# @Time : 2024/9/19 14:20
# @Author : wangchao124
# @Email : wangchao124@hikvision.com
# @File : SDKFunctionDemo.py
# @Software : PyCharm
import time

from ..Common import ConfRead, glo
from ..SdkService.CmsService import CmsDemo
from ..SdkService.AlarmService import AlarmDemo
from ..SdkService.SsService import SsDemo
from ..SdkService.StreamService import StreamDemo, VoiceDemo


def dispatch(command, lLoginID, cmsDemo, alarmDemo, ssDemo, streamDemo, VoiceDemo):
    if command == "f0001":
        print("\n[Function]CMS控制信令透传(需要设备在线, 获取NVR设备工作状态、远程升级等)")
        cmsDemo.CMS_XMLRemoteControl()

    elif command == "f0002":
        print("\n[Function]ISUP透传接口(获取云存储配置)")
        cmsDemo.CMS_ISAPIPassThrough("GET /ISAPI/System/PictureServer?format=json", "", "")

    elif command == "f0003":
        print("\n[Function]ISUP透传接口(获取设备信息)")
        cmsDemo.CMS_ISAPIPassThrough("GET /ISAPI/System/deviceInfo", "", "")

    elif command == "f0004":
        print("\n[Function]ISUP透传接口(获取设备校时参数)")
        cmsDemo.CMS_ISAPIPassThrough("GET /ISAPI/System/time", "", "")

    elif command == "f0005":
        print("\n[Function]ISUP透传接口(设置设备校时参数)")
        parameter = {"time": "2024-09-23T14:12:47+08:00"}  # 设备校时参数
        time_parameter = ConfRead.get_req_body_from_template("../../resources/conf/deviceParameter/time.xml", parameter)
        print(time_parameter)
        cmsDemo.CMS_ISAPIPassThrough("PUT /ISAPI/System/time", time_parameter, "")

    elif command == "f2000":
        print("\n[Function]上传图片至存储服务器(返回图片URL可以用于人脸下发)")
        pic_name = input(
            "请输入您想要进行图片分析的图片URL地址(图片URL可以参见 f2000:上传图片至存储服务器(返回图片URL可以用于人脸下发) 获取)")
        pic_path = "..//..//resources//pics//" + pic_name
        ssDemo.ssCreateClient(pic_path)
        ssDemo.ssUploadPic()
        ssDemo.ssDestroyClient()

    elif command == "f2001":
        print("\n[Function]抓图")
        cmsDemo.CMS_ISAPIPassThrough(
            "GET /ISAPI/Streaming/channels/101/picture/async?format=json&imageType=JPEG&URLType=cloudURL", "", "")

    elif command == "f3001":
        print("\n[Function]取流预览模块(有预览窗口)")


    elif command == "f3002":
        print("\n[Function]取流预览模块(保存到本地文件)")
        streamDemo.startRealPlayListen_File()
        streamDemo.RealPlay(channel=1)
        time.sleep(5)  # 程序等待 5 秒
        streamDemo.StopRealPlay()


    elif command == "f3003":
        print("\n[Function]按时间回放模块(有回放窗口)")


    elif command == "f3004":
        print("\n[Function]按时间回放模块(保存到本地文件)")
        # 回放监听函数调用
        streamDemo.startPlayBackListen_FILE()
        streamDemo.PlayBackByTime(lchannel=1)
        while not glo.get_value("stopPlayBackFlag"):
            try:
                time.sleep(1)  # 每秒检查一次
            except KeyboardInterrupt:
                # 如果线程被中断，通常需要清除中断状态
                print("Playback was interrupted")
                glo.set_value("stopPlayBackFlag", True)  # 设置标志位，退出循环
                break  # 当前线程被中断时退出循环
        streamDemo.stopPlayBackByTime()
        return

    elif command == "f3005":
        print("\n[Function]语音转发")
        VoiceDemo.start_voice_serve_listen()
        VoiceDemo.start_voice_trans()
        VoiceDemo.StopVoiceTrans()


    elif command == "f3006":
        print("\n[Function]ISUP透传接口(云台控制)")
        parameter = {"pan": "-60", "tilt": "0", "zoom": "0"}  # 向左转动
        in_parameter = ConfRead.get_req_body_from_template("../../resources/conf/video/PTZControl.xml", parameter)
        print(in_parameter.decode('utf-8'))
        cmsDemo.CMS_ISAPIPassThrough("PUT /ISAPI/PTZCtrl/channels/1/continuous", in_parameter, "")
        time.sleep(5)
        parameter = {"pan": "0", "tilt": "0", "zoom": "0"}  # 停止转动
        in_parameter = ConfRead.get_req_body_from_template("../../resources/conf/video/PTZControl.xml", parameter)
        print(in_parameter.decode('utf-8'))
        cmsDemo.CMS_ISAPIPassThrough("PUT /ISAPI/PTZCtrl/channels/1/continuous", in_parameter, "")
    elif command == "f3008":
        cmsDemo.FindRecored()


    else:
        print("未定义的指令类型!" + command)
